<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Download" data-toc-modified-id="Download-0.0.1"><span class="toc-item-num">0.0.1&nbsp;&nbsp;</span>Download</a></span></li></ul></li></ul></li><li><span><a href="#Open-CV" data-toc-modified-id="Open-CV-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Open CV</a></span><ul class="toc-item"><li><span><a href="#영상처리-개요" data-toc-modified-id="영상처리-개요-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>영상처리 개요</a></span></li><li><span><a href="#ROI" data-toc-modified-id="ROI-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>ROI</a></span></li><li><span><a href="#히스토그램" data-toc-modified-id="히스토그램-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>히스토그램</a></span><ul class="toc-item"><li><span><a href="#2D히스토그램-으로-색조를-나타냄" data-toc-modified-id="2D히스토그램-으로-색조를-나타냄-1.3.1"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>2D히스토그램 으로 색조를 나타냄</a></span></li></ul></li><li><span><a href="#Video" data-toc-modified-id="Video-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Video</a></span></li><li><span><a href="#그리기" data-toc-modified-id="그리기-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>그리기</a></span><ul class="toc-item"><li><span><a href="#선,-사각형,-원-그리기" data-toc-modified-id="선,-사각형,-원-그리기-1.5.1"><span class="toc-item-num">1.5.1&nbsp;&nbsp;</span>선, 사각형, 원 그리기</a></span></li><li><span><a href="#다각형" data-toc-modified-id="다각형-1.5.2"><span class="toc-item-num">1.5.2&nbsp;&nbsp;</span>다각형</a></span></li><li><span><a href="#문자열-표시하기" data-toc-modified-id="문자열-표시하기-1.5.3"><span class="toc-item-num">1.5.3&nbsp;&nbsp;</span>문자열 표시하기</a></span></li></ul></li><li><span><a href="#이벤트-처리" data-toc-modified-id="이벤트-처리-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>이벤트 처리</a></span><ul class="toc-item"><li><span><a href="#플래그를-이용한-동그라미-그리기" data-toc-modified-id="플래그를-이용한-동그라미-그리기-1.6.1"><span class="toc-item-num">1.6.1&nbsp;&nbsp;</span>플래그를 이용한 동그라미 그리기</a></span></li><li><span><a href="#트백바" data-toc-modified-id="트백바-1.6.2"><span class="toc-item-num">1.6.2&nbsp;&nbsp;</span>트백바</a></span></li></ul></li><li><span><a href="#이미지-프로세싱" data-toc-modified-id="이미지-프로세싱-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>이미지 프로세싱</a></span><ul class="toc-item"><li><span><a href="#THRESHOLD" data-toc-modified-id="THRESHOLD-1.7.1"><span class="toc-item-num">1.7.1&nbsp;&nbsp;</span>THRESHOLD</a></span></li><li><span><a href="#알파-블렌딩(blending_alpha)" data-toc-modified-id="알파-블렌딩(blending_alpha)-1.7.2"><span class="toc-item-num">1.7.2&nbsp;&nbsp;</span>알파 블렌딩(blending_alpha)</a></span></li><li><span><a href="#seamlessClone()" data-toc-modified-id="seamlessClone()-1.7.3"><span class="toc-item-num">1.7.3&nbsp;&nbsp;</span>seamlessClone()</a></span></li><li><span><a href="#CLAHE-적용" data-toc-modified-id="CLAHE-적용-1.7.4"><span class="toc-item-num">1.7.4&nbsp;&nbsp;</span>CLAHE 적용</a></span></li></ul></li><li><span><a href="#화질개선" data-toc-modified-id="화질개선-1.8"><span class="toc-item-num">1.8&nbsp;&nbsp;</span>화질개선</a></span><ul class="toc-item"><li><span><a href="#모폴로지-연산" data-toc-modified-id="모폴로지-연산-1.8.1"><span class="toc-item-num">1.8.1&nbsp;&nbsp;</span>모폴로지 연산</a></span></li><li><span><a href="#확대와-축소" data-toc-modified-id="확대와-축소-1.8.2"><span class="toc-item-num">1.8.2&nbsp;&nbsp;</span>확대와 축소</a></span></li><li><span><a href="#어핀-변환" data-toc-modified-id="어핀-변환-1.8.3"><span class="toc-item-num">1.8.3&nbsp;&nbsp;</span>어핀 변환</a></span></li><li><span><a href="#원근-변환" data-toc-modified-id="원근-변환-1.8.4"><span class="toc-item-num">1.8.4&nbsp;&nbsp;</span>원근 변환</a></span></li><li><span><a href="#마우스와-원근-변환으로-문서-스캔" data-toc-modified-id="마우스와-원근-변환으로-문서-스캔-1.8.5"><span class="toc-item-num">1.8.5&nbsp;&nbsp;</span>마우스와 원근 변환으로 문서 스캔</a></span><ul class="toc-item"><li><span><a href="#과제" data-toc-modified-id="과제-1.8.5.1"><span class="toc-item-num">1.8.5.1&nbsp;&nbsp;</span>과제</a></span></li></ul></li></ul></li></ul></li></ul></div>

### Download

다운로드
pip install openCV-python
pip install opencv-contrib-python  (다른 모듈도 같이 포함되어 있음)

In [ ]:

import cv2
fname = './images/lena.jpg'
original = cv2.imread(fname, cv2.IMREAD_COLOR)
cv2.imshow('Original', original)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [ ]:
 # numpy 배열로 반환 3 의 뜻은 3차원 배열로 color 
print(type(original), original.shape )

In [ ]:
import matplotlib.pyplot as plt

img =cv2.imread(fname, cv2.IMREAD_COLOR)
plt.imshow(img)
plt.show()

In [ ]:
b, g, r = cv2.split(img)  #오픈 CV의 채널 순서 :BGR
img2 = cv2.merge([r, g, b]) #numpy는 이미지를 RGB순으로 채널을 읽음
plt.imshow(img2)
plt.show()

In [ ]:
#pip install jupyter_contrib_nbextensions && jupyter contrib nbextension install

# Open CV
## 영상처리 개요

In [ ]:
import cv2
img = cv2.imread('./images/lena.jpg')

In [ ]:
img.dtype

In [ ]:
height, width, channel = img.shape  #컬러의 경우


In [ ]:
for y in range(height):
    for x in range(width):
        if(y%2 ==0):
            img[y,x] = 0

In [ ]:
import numpy as np

def black_lines(img):
    height, width = img.shape
    img_ = np.zeros(img.shape, dtype = np.uint8)
    for y in range(height):
        if(y%2 ==0):
            img_[y, :] =0
        else:
            img_[y, :] = img[y, :]
    return img_




In [ ]:
def img_pro(func, img, *args, show=True, **kwargs):
    img_= func(img, *args, **kwargs)
    if show:
        cv2.imshow("output", img_)
        cv2.waitKey()
        cv2.destroyAllWindows()
    else:
        return img_

def img_show(img, *args, show=True, **kwargs):
    if show:
        cv2.imshow("output", img)
        cv2.waitKey() #1밀리초안에 key값이 들어가야함 넘어가면 -1
        cv2.destroyAllWindows()
    else:
        return img_

In [ ]:
img = cv2.imread("./images/lena.jpg", flags = cv2.IMREAD_GRAYSCALE)
img_pro(black_lines, img)

## ROI

In [ ]:
import cv2

img = cv2.imread('./images/lena.jpg')

In [ ]:
temp = img[200:380, 200:360]
cv2.imshow('img', img)
cv2.imshow('roi_img', temp)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [ ]:
import numpy as np
b, g, r = cv2.split(img)
np.mean(b), np.mean(g), np.mean(r)

In [ ]:
img[:, :, [1,2]] = 0  # red링 green 0으로 
img_show(img)

In [ ]:
def weighted_blending(img1, img2, img1_weight=0.5):
    output = np.zeros(img1.shape, dtype = np.uint8)
    height, width = img1.shape[0 :2]
    if len(img1.shape) ==2:
        for y in range(height):
            for x in range(width):
                output[x, y] = int(img1[y,x]* img1_weight + img2[y, x]* (1-img1_weight))
    elif len(img1.shape) ==3:
        for y in range(height):
            for x in range(width):
                blended = int(img1[y,x]* img1_weight + img2[y, x]* (1-img1_weight))
                output[y, x] = blended.astype(np.uint8)
    
    return output

In [ ]:
def ez_blending(img1, img2, img1_weight=0.5):
    output = img1 * img1_weight + img2*(1-img1_weight)
    return output.astype(np.uint8)


In [ ]:
import numpy as np
img1 = cv2.imread('./images/flower1.jpg')
img2 = cv2.imread('./images/flower2.jpg')
new_img = ez_blending(img1, img2, img1_weight= 0.3)
img_show(new_img)

In [ ]:
img1 = cv2.imread('./images/logo.png')
img2 = cv2.imread('./images/flower2.jpg')

rows, cols, channels = img1.shape
roi = img2[0:rows, 0:cols]  #대상에서 삽입할 이미지의 영역을 추출

#mask로 만들기 위해서 img1을 gray로 변경후 binary image로 전환
#mask는 로고가 흰색(255), 배경이 검은색 (0)
#mask_inv 는 mask의 반대
img1_gray = cv2.cvtColor(img1, cv2.COLOR_BGR2GRAY)
ret, mask = cv2.threshold(img1_gray, 100,255, cv2.THRESH_BINARY)
mask_inv = cv2.bitwise_not(mask)

#and연산은 둘다 0이 아닌 경우만 통과시키며, mask에서 검정이 아닌 부분 로고부분이 나옴

img_fg = cv2.bitwise_and(img1, img1, mask=mask)
img2_bg = cv2.bitwise_and(roi, roi, mask= mask_inv)
dst = cv2.add(img_fg, img2_bg)

img2[0:rows, 0:cols] = dst

img_show(img2)

In [ ]:
img1 = cv2.imread('./images/logo.png')
img2 = cv2.imread('./images/flower2.jpg')

mask = np.full_like(img1, 255)
h, w = img2.shape[:2]
center = (w//2, h//2)
normal = cv2.seamlessClone(img1, img2, mask, center, cv2.NORMAL_CLONE) #모자이크
mixed  = cv2.seamlessClone(img1, img2, mask, center, cv2.MIXED_CLONE)  #투명

img_show(normal)
img_show(mixed)

In [ ]:
# True는 1, False 0
def two_tone(img, threshold= 128):
    output = (img > threshold) * 255
    return output.astype(np.uint8)

In [ ]:
img2 = cv2.imread('./images/flower2.jpg', 0)
new_img = two_tone(img2)
img_show(new_img)

In [ ]:
import matplotlib.pyplot as plt
hou = cv2.imread('./images/house-full.jpg', 0)
ret, th1= cv2.threshold(hou, 127, 255, cv2.THRESH_BINARY)

th2 = cv2.adaptiveThreshold(hou, 255, cv2.ADAPTIVE_THRESH_MEAN_C, cv2.THRESH_BINARY, 15, 2)
th3 = cv2.adaptiveThreshold(hou, 255, cv2.ADAPTIVE_THRESH_GAUSSIAN_C, cv2.THRESH_BINARY, 15, 2)

title = ['original', 'global', 'mean', 'gaussian']

images=[hou, th1, th2, th3]

for i in range(4):
    plt.figure(figsize=(10, 10))
    plt.subplot(2, 2, i+1), plt.imshow(images[i], 'gray')
    plt.title(title[i])
    plt.xticks([]), plt.yticks([])
    
plt.show()

## 히스토그램

In [ ]:
import cv2
import numpy as np
import matplotlib.pyplot as plt

img = cv2.imread('./images/couple2.jpg', cv2.IMREAD_GRAYSCALE)
hist = np.zeros((256))
h, w = img.shape
for y in range(h):
    for x in range(w):
        hist[img[y,x]] = hist[img[y,x]] +1
        
plt.bar(x=range(256), height=hist, width =1)

In [ ]:
lena_gray = cv2.imread("./images/lena.jpg", 1)
hist = cv2.calcHist(images=[lena_gray], channels=[0], mask=None, 
                   histSize=[256], ranges=[0,256])
plt.plot(hist.flatten())
plt.show()

In [ ]:
lena =cv2.imread('./images/lena.jpg')
hist = cv2.calcHist(images=[lena], channels=[2], mask=None, 
                   histSize=[256], ranges=[0, 256])
plt.plot(hist.flatten())
plt.show()

### 2D히스토그램 으로 색조를 나타냄

In [ ]:
img = cv2.imread('./images/home.jpg')
hsv = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)
hist = cv2.calcHist([hsv], [0,1], None, [180, 256], [0, 180, 0, 256])

#print(hist.shape)  #180, 256

plt.imshow(hist)
plt.show()

In [ ]:
print(hist.shape)

In [ ]:
import cv2
img = cv2.imread('./images/lena.jpg')
img_show(img)

In [ ]:
#영상의 점 1개는 1바이트
#grayscale은 0부터 255(포함)까지 단계로 색(회색조)를 표현
#컬러영상 - RGB 표현, HSV 표현, alpha(불투명도)값으로 투명도를 조절함

#이미지형식: jpg, png, gif, bmp, tiff
#00000010 00000001 00000011 -> 061101 0711 0612(0이 6개 1이 2개)


In [ ]:
gray = cv2.cvtColor(img, cv2.COLOR_RGB2GRAY)

In [ ]:
img_show(gray)

In [ ]:
cv2.imwrite('./images/lena_gray.jpg', gray)

## Video

In [ ]:
import cv2
cap = cv2.VideoCapture("./images/puppies.mp4") # 노트불의 경우 0번째는 노트북의 카메라 

In [ ]:
if cap.isOpened():
    delay = int(1000/cap.get(cv2.CAP_PROP_FPS)) #1초: 1000/ 초당 프레임수
    while True:
        ret, img = cap.read() # 프레임을 한장씩 가져옴 ret는 있으면 True 
        if ret:
            cv2.imshow("Puppy", img)
            if cv2.waitKey(delay) & 0xFF ==27: #esc누르면 꺼짐
                break
        else:
            break
else:
    print("비디오 안열림")
cap.release()
cv2.destroyAllWindows()

In [ ]:
if cap.isOpened():
    delay = int(1000/cap.get(cv2.CAP_PROP_FPS)) #1초: 1000/ 초당 프레임수 숫자가 작을수록 빨라짐
    while True:
        ret, img = cap.read() # 프레임을 한장씩 가져옴 ret는 있으면 True 
        if ret:
            gray = cv2.cvtColor(img, cv2.COLOR_RGB2GRAY)
            cv2.imshow("Puppy", gray)
            if cv2.waitKey(delay) & 0xFF ==27: #esc누르면 꺼짐
                break
        else:
            break
else:
    print("비디오 안열림")
cap.release() #release를 해야 비디오가 나옴
cv2.destroyAllWindows()

In [ ]:
fourcc = cv2.VideoWriter_fourcc(*'DIVX') #DIVX, XVID, FMP4, X264, MJPG
width = cap.get(cv2.CAP_PROP_FRAME_WIDTH)
height = cap.get(cv2.CAP_PROP_FRAME_HEIGHT)
size = (int(width), int(height))
fps = cap.get(cv2.CAP_PROP_FPS)

out = cv2.VideoWriter("video.avi", fourcc, fps, size)
if cap.isOpened():
    delay = int(1000/cap.get(cv2.CAP_PROP_FPS)) #1초: 1000/ 초당 프레임수 숫자가 작을수록 빨라짐
    while True:
        ret, img = cap.read() # 프레임을 한장씩 가져옴 ret는 있으면 True 
        if ret:
            gray = cv2.cvtColor(img, cv2.COLOR_RGB2GRAY)
            cv2.imshow("Puppy", gray)
            if cv2.waitKey(delay) & 0xFF ==27: #esc누르면 꺼짐
                print("윈도우 종료")
                break
        else:
            print(ret, img)
            break
else:
    print("비디오 안열림")
out.release()
cap.release() #release를 해야 비디오가 나옴
cv2.destroyAllWindows()


## 그리기

### 선, 사각형, 원 그리기

In [ ]:
#cv2.line(img, start, end, color, thickness=1)
#cv2.rectangle(img, start, end, color, thickness=1)
# thickness = -1이면 색이 채워진다.
import cv2
import numpy as np

img  = np.full((250, 250, 3), 255, dtype=np.uint8)

cv2.line(img, (50, 10), (200, 10), (255, 0, 0), 1)
cv2.line(img, (50, 60), (200, 60), (255, 256, 0), 2)
cv2.line(img, (50, 110), (200, 110), (255, 0, 255), 3)
cv2.line(img, (50, 160), (200, 160), (0, 0, 255), 4)

cv2.imshow('Lines', img)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [ ]:
img  = np.full((250, 250, 3), 255, dtype=np.uint8)

cv2.rectangle(img, (50, 10), (200, 50), (255, 0, 0), 1)
cv2.rectangle(img, (50, 60), (200, 100), (255, 256, 0), 5)
cv2.rectangle(img, (50, 110), (200, 150), (255, 0, 255), -1)
cv2.rectangle(img, (50, 160), (200, 200), (0, 0, 255), 5)
cv2.rectangle(img, (50, 160), (200, 200), (0, 255, 255), -1)

cv2.imshow('Rectangle', img)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [ ]:
#cv2.circle(img, center, redius, color, thickness=1)

img = np.full((250, 250, 3), 255, dtype=np.uint8)

cv2.circle(img, (50, 50), 20, (255, 0, 0))
cv2.circle(img, (150, 50), 50, (255, 255, 0), -1)
cv2.circle(img, (200, 150), 30, (0, 0, 0), -1)
cv2.circle(img, (150, 50), 100, (255, 255, 0), 2)

cv2.imshow('Circle', img)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [ ]:
#cv2.ellipse(img, center, axes, angle, startAngle, endAngle, color, thickness=1)

img = np.full((250, 250, 3), 255, dtype=np.uint8)

cv2.ellipse(img, (50, 50),(50,30), -30, 0, 360, (255, 0, 0))
cv2.circle(img, (150, 50), 50, (255, 255, 0), -1)
cv2.circle(img, (200, 150), 30, (0, 0, 0), -1)
cv2.circle(img, (150, 50), 100, (255, 255, 0), 2)

cv2.imshow('Ellipse', img)
cv2.waitKey(0)
cv2.destroyAllWindows()

### 다각형

In [ ]:
# cv2.polylines(img, points, isClosed, color, thickness)

img = np.full((250, 250, 3), 25, dtype=np.uint8)
points = np.array([[[60, 60], [100, 10], [200, 60], [150, 100], [100, 100]]], dtype=np.int32)

cv2.polylines(img, points, True, (255, 0, 0), 5)
cv2.polylines(img, points+[0, 100], False, (0, 0, 255), 10)

cv2.imshow('Polylines', img)
cv2.waitKey(0)
cv2.destroyAllWindows()

### 문자열 표시하기

In [ ]:
#cv2.putText(img, text, org, font, fontscale, color)
# org = 글자의 왼쪽 아래의 좌표

img = np.full((250, 250, 3), 255, dtype=np.uint8)

cv2.putText(img, "HELLO", (10, 50), cv2.FONT_ITALIC, 1, (255,0,0))
cv2.putText(img, "Python", (50, 150), cv2.FONT_ITALIC, 1, (255,0,0))

cv2.imshow('Txt', img)
cv2.waitKey(0)
cv2.destroyAllWindows()

## 이벤트 처리

### 플래그를 이용한 동그라미 그리기

In [1]:


import cv2

title = 'mouse event'
img = cv2.imread('./images/hand.png')
cv2.imshow(title, img)

colors = {'black': (0, 0, 0),
        'red': (0, 0, 255),
        'green': (0, 255, 0),
        'blue': (255, 0, 0)}
def onMouse(event, x, y, flags, param):
    print(event, x, y, flags)
    color = colors['black']
    if event == cv2.EVENT_FLAG_LBUTTON:
        #컨트롤키와 시프트키를 모두 누른 경우
        if flags & cv2.EVENT_FLAG_SHIFTKEY and flags & cv2.EVENT_FLAG_CTRLKEY:
            color = colors['green']
        elif flags & cv2.EVENT_FLAG_SHIFTKEY: 
            color = colors['blue']
        elif flags & cv2.EVENT_FLAG_CTRLKEY:
            color = colors['red']
        cv2.circle(img, (x,y), 30, color, -1)
        cv2.imshow(title, img)
cv2.setMouseCallback(title, onMouse)  #마우스 콜백 함수를 GUI 윈도에 등록

while True:
    if cv2.waitKey(-1) & 0xFF == 27:
        break
cv2.destroyAllWindows()

1 185 148 1
0 185 148 1
0 185 147 1
0 185 147 1
4 185 147 1
0 102 94 16
0 102 94 16
0 102 94 16
0 102 94 16
0 100 97 16
0 99 98 16
0 96 100 16
0 96 100 16
0 96 101 16
0 96 101 16
0 96 101 16
0 95 101 16
0 95 101 16
0 95 101 16
0 95 101 0
0 94 101 0
0 93 100 0
0 92 100 0
0 90 98 0
0 89 98 0
0 89 98 0
0 88 97 0
0 85 95 0
0 83 93 0
0 81 92 0
0 81 91 0
0 80 90 16
0 79 89 16
0 79 89 16
0 79 89 16
0 79 89 16
1 79 89 17
0 79 89 17
4 79 89 17
0 78 89 16
0 78 89 16
0 78 89 16
0 78 89 16
0 77 90 16
0 77 90 16
0 77 90 16
0 77 90 16
0 77 90 16
0 76 90 16
0 73 91 16
0 69 91 16
0 65 92 16
0 57 93 16
0 52 93 16
0 45 93 16
0 44 93 16
0 41 93 16
0 41 93 16
0 40 91 16
0 39 89 16
0 38 85 16
0 37 84 16
0 37 80 16
0 36 76 16
0 36 75 16
0 35 72 16
0 35 67 16
0 35 63 16
0 36 60 16
0 39 54 16
0 41 48 16
0 44 42 16
0 48 36 16
0 49 35 16
0 52 32 16
0 54 29 16
0 54 29 16
0 54 29 16
0 54 29 16
0 55 29 16
1 55 29 17
4 55 29 17
0 55 29 16
0 55 29 16
0 56 29 16
0 58 29 16
0 61 29 16
0 62 29 16
0 65 29 16
0 69 30 16


0 183 90 24
0 185 97 24
0 186 98 24
0 187 102 24
0 189 106 24
0 191 109 24
0 193 114 24
0 194 115 24
0 197 122 24
0 199 125 24
0 200 127 24
0 200 127 24
0 201 128 24
0 201 129 24
0 201 130 24
0 202 130 24
0 202 131 24
0 202 131 24
0 202 131 24
0 202 132 24
0 202 132 24
0 202 132 24
0 202 133 24
0 202 133 24
0 203 133 24
0 203 133 24
0 203 133 24
0 203 133 24
0 203 134 24
0 203 134 24
0 203 134 24
0 203 134 24
0 203 134 8
0 203 134 8
0 203 133 8
0 203 133 8
0 203 133 8
0 203 132 0
0 202 131 0
0 202 130 0
0 202 129 0
0 202 129 0
0 201 128 0
1 201 128 1
4 201 128 1
0 202 129 0
0 202 129 0
0 205 132 0
0 207 134 0
0 208 135 0
0 209 136 0
0 209 136 0
0 210 137 0
0 210 137 0
1 210 137 1
4 210 137 1
0 212 140 0
0 215 143 0
0 218 146 0
0 221 148 0
0 223 149 0
0 226 152 0
0 227 152 0
0 227 152 0
1 227 152 1
4 227 152 1
0 227 152 0
0 227 153 0
0 228 154 0
0 228 155 0
0 228 156 0
0 228 156 0
0 228 157 0
0 228 158 0
0 229 158 0
0 229 158 0
0 229 159 0
0 229 159 0
0 230 159 0
0 230 158 0
0 230 158 0

0 185 158 0
0 185 159 0
0 186 159 0
0 186 159 0
0 186 159 0
0 186 159 0
0 186 160 16
0 187 160 16
0 187 160 16
0 187 160 16
0 187 160 16
0 187 161 16
0 187 161 16
0 188 161 16
0 188 161 16
0 188 161 16
0 188 162 16
0 189 162 16
0 189 162 16
0 190 162 0
0 190 162 0
0 190 162 0
0 190 162 0
0 190 163 0
1 190 163 9
4 190 163 9
0 191 163 8
0 191 163 8
0 192 163 8
0 193 164 8
0 195 164 8
0 196 164 8
0 198 164 8
0 199 164 8
0 199 164 8
0 200 165 0
0 200 165 0
0 201 165 0
0 201 165 0
0 201 165 0
0 201 165 0
0 202 165 0
0 202 165 0
0 202 165 0
0 202 165 0
0 203 165 0
0 203 165 0
0 205 165 0
0 208 165 0
0 209 165 0
0 213 164 0
0 217 164 0
0 220 163 0
0 227 163 0
0 232 163 0
0 235 162 0
0 238 162 0
0 240 162 0
0 243 162 32
0 244 162 32
0 244 162 32
0 245 162 32
0 245 162 32
1 245 162 33
4 245 162 33
0 245 162 32
0 245 162 32
0 245 162 0
0 244 162 0
0 244 163 0
0 244 163 0
0 244 163 0
0 244 163 0
0 242 164 0
0 240 164 0
0 238 164 0
0 236 164 0
0 236 164 0
0 235 164 0
0 234 164 0
0 234 164 0
0 234 

### 트백바

In [ ]:
import cv2
import numpy as np

win_name = 'Trackbar'

img  = np.full((500, 500, 3), 255, dtype=np.uint8)
cv2.imshow(win_name, img)

def onChange(x):
    print(x)
    # R G B 의 트백바 위치값
    r = cv2.getTrackbarPos('R', win_name)
    g = cv2.getTrackbarPos('G', win_name)
    b = cv2.getTrackbarPos('B', win_name)
    
    print(r, g, b)
    img[:] = [b, g, r]
    cv2.imshow(win_name, img)
    
#트랙바 생성
cv2.createTrackbar('R', win_name, 255, 255, onChange)
cv2.createTrackbar('G', win_name, 255, 255, onChange)
cv2.createTrackbar('B', win_name, 255, 255, onChange)

while True:
    if cv2.waitKey(-1) & 0xFF ==27:
        break
cv2.destroyAllWindows()

#141, 137, 130, 126
## 이미지 프로세싱 
### THRESHOLD
 - 이미지를 여러 영역으로 나눈 다음에 그 주변 픽셀 값만 가지고 계싼을 해서 경계 값을 구하는 방법 (적응형 스레시 홀드)
 

In [ ]:
import cv2
import numpy as np
import matplotlib.pylab as plt

img = cv2.imread('./images/gradient.png', cv2.IMREAD_GRAYSCALE)


_, t_bin = cv2.threshold(img, 127, 255, cv2.THRESH_BINARY)
_, t_bininv = cv2.threshold(img, 127, 255, cv2.THRESH_BINARY_INV)
_, t_truc = cv2.threshold(img, 127, 255, cv2.THRESH_TRUNC)
_, t_2zr = cv2.threshold(img, 127, 255, cv2.THRESH_TOZERO)
_, t_2zrinv = cv2.threshold(img, 127, 255, cv2.THRESH_TOZERO_INV)


imgs = {'original': img, 'BINARY': t_bin, 'BINART_INV': t_bininv, 'TRUNC':t_truc, 'TOZERO':t_2zr, 'TOZEROINV': t_2zrinv}

for i, (key, value) in enumerate(imgs.items()):
    plt.subplot(2,3, i+1)
    plt.title(key)
    plt.imshow(value, cmap='gray')
    plt.xticks([]); plt.yticks([])

plt.show()


In [ ]:
img = cv2.imread('./images/dave.png', cv2.IMREAD_GRAYSCALE)

blk_size=9 #블록 사이즈
C=5 # 차감상수

#오츠의 알고리즘으로 단일 경계값을 전체 이미지에 적용
ret, th1 = cv2.threshold(img, 0, 255, cv2.THRESH_BINARY | cv2.THRESH_OTSU)

# 적응형 스레시 홀드를 평균과 가우시안 분포로 각각 적용

th2 = cv2.adaptiveThreshold(img, 255, cv2.ADAPTIVE_THRESH_MEAN_C, cv2.THRESH_BINARY, blk_size, C)
th3 = cv2.adaptiveThreshold(img, 255, cv2.ADAPTIVE_THRESH_GAUSSIAN_C, cv2.THRESH_BINARY, blk_size, C)

imgs = {'Original': img, 'Global-Otsu:%d'%ret:th1, 'Adapted-Mean':th2, 'Adapted-Gaussian':th3}
for i, (k,v) in enumerate(imgs.items()):
    plt.subplot(2,3, i+1)
    plt.title(k)
    plt.imshow(v, cmap='gray')
    plt.xticks([]); plt.yticks([])

plt.show()

### 알파 블렌딩(blending_alpha)

In [ ]:
import cv2
import numpy as np
alpha = 0.8

#합성에 사용할 영상
img1 = cv2.imread('./images/tes.jpg')
img1 = cv2.resize(img1, (500, 500))
img2 = cv2.imread('./images/flowerback.jpg')
img2 = cv2.resize(img2, (500, 500))

#수식을 직접 연산해서 알파 블렌딩 적용
blended = img1*alpha + img2*(1-alpha)
blended = blended.astype(np.uint8)  #소수점 제거
cv2.imshow('img1*alpha + img2*(1-alpha)', blended)

#addWeighted() 함수로 알파 블렌딩 적용
dst = cv2.addWeighted(img1, alpha, img2, (1-alpha), 0)
cv2.imshow('addWeighted', dst)

cv2.waitKey(0)
cv2.destroyAllWindows()

In [ ]:
import numpy as np, cv2
import matplotlib.pylab as plt

img = cv2.imread('./images/hip.png')
img = cv2.resize(img, (500, 500))

mask = np.zeros_like(img)
cv2.circle(mask, (250, 100), 200, (255, 255, 255), -1)
#cv2.circle(대상 이미지, (원점x, 원점y), 반지름, (색상), 채우기)

#마스킹
masked = cv2.bitwise_and(img, mask)

cv2.imshow('original', img)
cv2.imshow('mask', mask)
cv2.imshow('masked', masked)

cv2.waitKey()
cv2.destroyAllWindows()

### seamlessClone()

In [ ]:
import cv2
import numpy as np

src = cv2.imread('./images/plane.jpg') #w전경
dst = cv2.imread('./images/back.jpg') # 배경
h, w = dst.shape[:2]

src_mask = np.zeros(src.shape, src.dtype)
poly = np.array([[0, 40], [150, 60], [203, 56], [200, 100], [88, 103], [1, 75]], np.int32)

cv2.fillPoly(src_mask, [poly], (255,255,255))

center = (500, 100)
output = cv2.seamlessClone(src, dst, src_mask, center, cv2.MIXED_CLONE)

cv2.imshow('output', output)
cv2.waitKey()
cv2.destroyAllWindows()

In [ ]:
import cv2
import numpy as np
import matplotlib.pylab as plt

img = cv2.imread('./images/sunset.jpg', cv2.IMREAD_GRAYSCALE)
rows, cols = img.shape[:2]
hist = cv2.calcHist([img], [0], None, [256], [0, 256])
cdf = hist.cumsum()
cdf_m = np.ma.masked_equal(cdf, 0)
cdf_m = (cdf_m - cdf_m.min())/ (rows * cols) *255 # 이퀄라이즈 히스토그램 계산 
cdf = np.ma.filled(cdf_m, 0).astype('uint8') #NaN을 0으로
img2 = cdf[img] # 히스토그램 픽셀로 매핑

#openCV API로 이퀄라이즈 히스토그램 적용
img3 = cv2.equalizeHist(img)

hist2=cv2.calcHist([img2], [0], None, [256], [0, 256])
hist3=cv2.calcHist([img3], [0], None, [256], [0, 256])

cv2.imshow('Before', img)
cv2.imshow('Manual', img2)
cv2.imshow('equalize', img3)

hists= {'Befoer': hist, 'Manual': hist2, 'equalize': hist3}
for i, (k,v) in enumerate(hists.items()):
    plt.subplot(2, 3, i+1)
    plt.title(k)
    plt.plot(v)
plt.show()

### CLAHE 적용

In [ ]:
img = cv2.imread('./images/butterfly.jpg')
img_yu = cv2.cvtColor(img, cv2.COLOR_BGR2YUV)

#밝기 채널에 대해서 이퀄라이즈 적용
img_eq = img_yu.copy()
img_eq[:,:,0] = cv2.equalizeHist(img_eq[:,:,0])
img_eq = cv2.cvtColor(img_eq, cv2.COLOR_YUV2RGB)

#밝기 채널에 대해서 CLAHE 적용
img_clahe = img_yu.copy()
clahe = cv2.createCLAHE(clipLimit = 3.0, tileGridSize=(8,8))
img_clahe[:, :, 0] = clahe.apply(img_clahe[:,:,0])
img_clahe = cv2.cvtColor(img_clahe, cv2.COLOR_YUV2BGR)

#결과 출력 
cv2.imshow('Before', img)
cv2.imshow('CLAHE', img_clahe)
cv2.imshow('equalizeHist', img_eq)
cv2.waitKey(0)
cv2.destroyAllWindows()

## 화질개선
### 모폴로지 연산
 - 침색, 팽창, 열기, 닫기
 181, 186, 188, 190 
 ### 확대와 축소


In [ ]:
import cv2
import numpy as np

img = cv2.imread('./images/hip.png')
img = cv2.resize(img, (500, 500))

h, w = img.shape[:2]

dst1 = cv2.resize(img, (int(w*0.5), int(h*0.5)), interpolation=cv2.INTER_AREA)

# 배율 지정으로 확대

dst2 = cv2.resize(img, None, None, 2, 2, cv2.INTER_CUBIC)

cv2.imshow("original", img)
cv2.imshow("small", dst1)
cv2.imshow("big", dst2)
cv2.waitKey(0)
cv2.destroyAllWindows()

### 어핀 변환

In [ ]:
import matplotlib.pylab as plt

fname = './images/hip.png'
img = cv2.imread(fname)
img = cv2.resize(img, (500,500))

pts1 = np.float32([[100,50], [200, 50], [100, 200]])
pts2 = np.float32([[80, 70], [210, 60], [250, 120]])

#변환전 좌표 표시
cv2.circle(img, (100, 50), 5, (255, 0), -1)
cv2.circle(img, (200, 50), 5, (0, 255, 0), -1)
cv2.circle(img, (100, 200), 5, (0, 0, 255), -1)

#짝지은 3개의 좌표로 변환행렬 계싼
mtrx = cv2.getAffineTransform(pts1, pts2)

#어핀 변환 적용
dst = cv2.warpAffine(img, mtrx, (int(cols*1.5), rows))


#결과 출력
cv2.imshow('origin', img)
cv2.imshow('affin', dst)
cv2.waitKey(0)
cv2.destroyAllWindows()

### 원근 변환

In [ ]:
import cv2
import numpy as np

fname = './images/hip.png'
img = cv2.imread(fname)
img = cv2.resize(img, (500, 500))

rows, cols = img.shape[:2]

pts1 = np.float32([[0,0], [0, rows], [cols,0], [cols, rows]])
pts2 = np.float32([[100, 50], [10, rows-50], [cols-100, 50], [cols-10, rows-50]])

#변환전 원본 이미지에 좌표 포시
cv2.circle(img, (0, rows), 5, (255, 0), -1)
cv2.circle(img, (cols, 0), 5, (0, 255, 0), -1)
cv2.circle(img, (0, 0), 5, (0, 0, 255), -1)
cv2.circle(img, (cols, rows), 5, (0, 255, 255), -1)

#원근 변환행렬 계산
mtrx = cv2.getPerspectiveTransform(pts1, pts2)

#원근 변환 적용
dst = cv2.warpPerspective(img, mtrx, (cols, rows))

cv2.imshow('origin', img)
cv2.imshow('perspective', dst)
cv2.waitKey(0)
cv2.destroyAllWindows()

### 마우스와 원근 변환으로 문서 스캔

In [ ]:
import cv2
import numpy as np

fname = './images/hip.png'
win_name = 'scanning'
img = cv2.imread(fname)
img = cv2.resize(img, (500, 740))

rows, cols = img.shape[:2]
draw = img.copy()
pts_cnt = 0
pts = np.zeros((4,2), dtype=np.float32)

def onMouse(event, x, y, flags, param):
    global pts_cnt
    if event == cv2.EVENT_LBUTTONDOWN:
        cv2.circle(draw, (x,y), 10, (0, 255, 0), -1)
        cv2.imshow(win_name, draw)
        pts[pts_cnt] = [x,y]
        pts_cnt +=1
        if pts_cnt ==4:
            #좌표 4개 중 상하좌우 찾기
            sm = pts.sum(axis=1)
            diff = np.diff(pts, axis = 1)
            topleft = pts[np.argmin(sm)]
            bottom = pts[np.argmax(sm)]
            topright = pts[np.argmin(diff)]
            bottomLeft = pts[np.argmax(diff)]
            
            pts1 = np.float32([topleft, topright, bottom, bottomLeft])
            
            w1 = abs(bottom[0]-bottomLeft[0])
            w2 = abs(topright[0]-topleft[0])
            h1 = abs(topright[1]-bottom[1])
            h2 = abs(topleft[1]- bottomLeft[1])
            
            width = max([w1, w2])
            height = max([h1, h2])
             
            pts2 = np.float32([[0,0], [width-1, 0], [width-1, height-1], [0, height-1] ])
            
            mtrx = cv2.getPerspectiveTransform(pts1, pts2)
            result = cv2.warpPerspective(img, mtrx, (width, height))
            cv2.imshow('scann', result)
            
cv2.imshow(win_name, img)
cv2.setMouseCallback(win_name, onMouse)
cv2.waitKey(0)
cv2.destroyAllWindows()

#### 과제

In [ ]:
import cv2
import numpy as np

win_name = 'scanning'
pts_cnt = 0
pts = np.zeros((4,2), dtype=np.float32)

def mouseHandler(event, x, y, flags, param):
    global pts_cnt
    if event == cv2.EVENT_LBUTTONDOWN:
        cv2.circle(img, (x,y), 10, (0, 255, 0), -1)
        cv2.imshow(win_name, img)
        pts[pts_cnt] = [x,y]
        pts_cnt +=1
        if pts_cnt ==4:
            #좌표 4개 중 상하좌우 찾기
            sm = pts.sum(axis=1)
            diff = np.diff(pts, axis = 1)
            topleft = pts[np.argmin(sm)]
            bottom = pts[np.argmax(sm)]
            topright = pts[np.argmin(diff)]
            bottomLeft = pts[np.argmax(diff)]
            
            pts1 = np.float32([topleft, topright, bottom, bottomLeft])
            
            w1 = abs(bottom[0]-bottomLeft[0])
            w2 = abs(topright[0]-topleft[0])
            h1 = abs(topright[1]-bottom[1])
            h2 = abs(topleft[1]- bottomLeft[1])
            
            width = max([w1, w2])
            height = max([h1, h2])
            
            pts2 = np.float32([[0,0], [width-1, 0], [width-1, height-1], [0, height-1] ])
            
            mtrx = cv2.getPerspectiveTransform(pts1, pts2)
            result = cv2.warpPerspective(img, mtrx, (width, height))
            cv2.imshow('scann', result)
    

cap = cv2.VideoCapture(0)
captured = False

if cap.isOpened():
    delay = int(1000/cap.get(cv2.CAP_PROP_FPS)) 
    while True:
        ret, img = cap.read() 
        if ret:
            cv2.imshow("Capture", img)
            key = cv2.waitKey(delay)
            if key & 0xFF ==27:
                print("아무 작업도 하지 않고 종료함")
                break
            elif key == ord('c'):
                captured = True
                break
        else:
            break
else:
    print("비디오 안열림")
    
if captured:
    cap.release()
    while True:
        cv2.imshow('Capture', img)
        cv2.setMouseCallback('Capture', mouseHandler)
        key = cv2.waitKey(delay)
        if key & 0xFF ==27:
            print("ESC")
            break
    
cap.release()
cv2.destroyAllWindows()